src/reference/packing.cc

```diff
@@ -1469,9 +1470,111 @@ void xnn_pack_qs4_weights_and_biases(
       extra_data1_element_size, packed_weights_ptr, extra_bytes, params);
 }
 
+size_t xnn_packed_stride_qb4_weights_and_biases(
+    const struct xnn_gemm_config* gemm_config, size_t k, size_t block_size,
+    size_t k_stride, size_t extra_bytes) {
+  const uint32_t kr = UINT32_C(1) << gemm_config->log2_kr;
+  const uint32_t sr = UINT32_C(1) << gemm_config->log2_sr;
+  const uint32_t nr = gemm_config->nr;
+  const size_t planes = gemm_config->planes;
+  
+  size_t input_channels = round_up_po2(k, planes);
+
+  size_t block_scale_bytes = 0;
+  size_t num_blocks = 0;
+  const bool block_wise = (block_size != 0);
+  if (block_wise) {
+    num_blocks = input_channels / block_size;
+    block_scale_bytes += num_blocks * sizeof(uint16_t);
+  }
+
+  const size_t bias_element_size = sizeof(int32_t);
+  const size_t log2_filter_element_size = XNN_LOG2_SIZEOF_INT8_T;
+  return (k_stride << log2_filter_element_size) + bias_element_size +
+         extra_bytes + block_scale_bytes;
+}
+
+void xnn_pack_qb4_weights_and_biases(
+    uint32_t flags, const struct xnn_gemm_config* gemm_config,
+    size_t input_channels, size_t output_channels, size_t groups,
+    size_t block_size, size_t k_stride, const void* accumulator_init, const void* weights,
+    xnn_init_scale_params_fn init_extra_data0_fn, const void* extra_data0,
+    size_t extra_data0_element_size,
+    xnn_init_scale_params_fn init_extra_data1_fn, const void* extra_data1,
+    size_t extra_data1_element_size, void* packed_weights_ptr,
+    const void* params) {
+
+  const uint32_t nr = gemm_config->nr;
+  const uint32_t kr = UINT32_C(1) << gemm_config->log2_kr;
+  const uint32_t sr = UINT32_C(1) << gemm_config->log2_sr;
+  const size_t planes = gemm_config->planes;
+
+  const size_t extra_bytes_bl = sizeof(uint16_t);
+  const size_t extra_bytes_n = sizeof(uint32_t);
+  if (flags & XNN_FLAG_TRANSPOSE_WEIGHTS) {
+    xnn_pack_qs8_qb4w_gemm_gio_w(
+      /*g=*/groups, 
+      /*nc=*/output_channels, 
+      /*kc=*/input_channels,
+      /*nr=*/nr, 
+      /*kr=*/kr, 
+      /*sr=*/sr,
+      /*k_stride=*/k_stride,
+      /*bl=*/block_size,
+      /*k=*/(const uint8_t*)weights, 
+      /*bias=*/NULL, 
+      /*scale=*/(const xnn_bfloat16*)extra_data1,
+      /*packed_weights=*/packed_weights_ptr,
+      /*extra_bytes_bl=*/nr * extra_bytes_bl,
+      /*extra_bytes_n=*/nr * extra_bytes_n,
+      /*params*/(const struct xnn_qs8_qc4w_packing_params *)params);
+  } else {
+    xnn_pack_qs8_qb4w_gemm_goi_w(
```

Comment:
can these use gemm-config to set the packw microkernels?

Reply:
do you mean using something like:
```
gemm_config->packw_gemm_goi(
```

I was contemplating this but i was looking at config-types.h:

https://github.com/google/XNNPACK/blob/c5354784c1ab1c999a14b199b9dbaeba4bd523bc/src/xnnpack/config-types.h#L189-L197

and there seem to be active tasks to instead move all of these into pack_weights_and_biases instead. So I went ahead and helped towards this by removing packw_gemm_goi_bl from the config types. Not sure what the intended direction for this is though. Perhaps @alankelly can chime in.

Reply:
This is the correct way of calling packing functions. Thanks for the clean up, this is much better

Reply:
i was discussing with @fbarchard and I think we might want to keep the 
```
xnn_packw_gemm_goi_bl_ukernel_fn pack_gemm_goi_bl; 
```

in the config-types so that we can specify the non reference packing ukernels in gemm-config, and the pack_weights_and_biases will pull the actual packing ukernel from the gemm-config. So in a sense pack_weights_and_biases will be the wrapper around packw_gemm_goi_ukernel, and will also fill in scales, biases at the same time

Reply:
I'm not sure where those deprecated comments came from but my main concern is that the pack_weights_and_biases is hard coded to the reference packing functions, preventing them from being removed in favor of scalar microkernels.

If we can make them call the configured microkernels, performance would be improved.

Secondary concern is that GIO packing is a transpose compared to GOI.  Unless there is a usecase, the GIO packing functions will likely be less optimized scalar kernels.

I've noticed packing signature mismatches.  I think this occurred when params were changed in packing.cc but the packw microkernels were not updated to the same signature.
They should match.  Whatever signature packing.cc has, the packw should be idential, and the function pointers, tests and benchmarks should be able to use the microkernels.